stdlib/camlinternalLazy.ml

```diff
@@ -50,43 +83,37 @@ let do_force_val_block blk =
   let closure = (Obj.obj (Obj.field b 0) : unit -> 'arg) in
   Obj.set_field b 0 (Obj.repr (domain_self ()));
   let result = closure () in
-  make_forward b (Obj.repr result);
+  Obj.set_field b 0 (Obj.repr result);
+  update_to_forward b;
   result
 
-type status = Racy | Forcing
-
-(* Assumes [blk] is a block with tag lazy *)
-let update_tag_forcing (blk : 'arg lazy_t) =
-  let b = Obj.repr blk in
-  if not (update_tag b Obj.lazy_tag Obj.forcing_tag) then
-    (* blk has tag either
-        + Obj.forcing_tag -- currently being forced by this domain or
-                             another concurrent domain (or)
-        + Obj.forward_tag -- was being forced by another domain which has since
-                             completed the evaluation and updated the lazy. *)
-    let forcing_domain_id : int = Obj.obj (Obj.field b 0) in
-    let my_domain_id = domain_self () in
-    (* XXX KC: Need a fence here to prevent the tag read from being reordered
-     * before reading the first field of [b] *)
-    if Obj.tag b = Obj.forcing_tag && forcing_domain_id = my_domain_id then
-      raise Undefined
-    else Racy
-  else Forcing
-
-let force_gen_lazy_block ~only_val blk =
-  match update_tag_forcing blk with
-  | Racy -> raise Undefined
-  | Forcing when only_val -> do_force_val_block blk
-  | Forcing -> do_force_block blk
+let force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
+  match update_to_forcing (Obj.repr blk) with
+  | 0 when only_val -> do_force_val_block blk
+  | 0 -> do_force_block blk
+  | 1 -> raise Undefined
+  | _ ->
+      (* We raise [RacyLazy] for all the other cases.
+       *
+       * If the return value was 2, we should raise RacyLazy. We had observed
+       * earlier in [force_gen] that the [blk] was not [Obj.forward_tag] and
+       * not short-circuited. But now we observe that the value has either been
+       * forwarded (return value 3) or forwarded and short-circuited (4). Both
+       * of these indicate a race. *)
+      raise RacyLazy
 
 (* used in the %lazy_force primitive *)
 let force_lazy_block blk = force_gen_lazy_block ~only_val:false blk
 
-let try_force_gen_lazy_block ~only_val blk =
-  match update_tag_forcing blk with
-  | Racy -> None
-  | Forcing when only_val -> Some (do_force_val_block blk)
-  | Forcing -> Some (do_force_block blk)
+let try_force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
+  match update_to_forcing (Obj.repr blk) with
```

Comment:
I suspect it's possible to run into the same error we were previously witnessing here.